specgen/specgen/hellocap/listener.go

```diff
@@ -0,0 +1,200 @@
+package hellocap
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+)
+
+const tlsRecordHeaderLen = 5
+
+type contextKey int
+
+// contextHelloKey is the context key for retrieving the ClientHello.
+const contextHelloKey contextKey = 0
+
+// FromRequest returns the TLS ClientHello sent by the remote. This request must have come through a
+// handler attached to the Server type in this package.
+func FromRequest(req *http.Request) ([]byte, error) {
+	hello, ok := req.Context().Value(contextHelloKey).(capturedHello)
+	if !ok {
+		return nil, errors.New("no attached ClientHello")
+	}
+	return hello.hello, hello.err
+}
+
+// attachHello is used to put a captured ClientHello on a http.Request context. This is how the
+// ClientHello, captured at the TCP level, is made available to users of this library.
+func attachHello(req *http.Request, hello capturedHello) *http.Request {
+	ctx := context.WithValue(req.Context(), contextHelloKey, hello)
+	return req.Clone(ctx)
+}
+
+// onHello is a callback invoked when a ClientHello is captured. Only one of hello or err will be
+// non-nil. This callback is not invoked for incomplete ClientHellos. In other words, err is non-nil
+// only if what has been read off the connection could not possibly constitute a ClientHello,
+// regardless of further data from the connection.
+type onHello func(hello []byte, err error)
+
+// capturingConn is a TCP connection and implements net.Conn. If a TLS ClientHello is sent by the
+// peer, capturingConn will invoke the provided onHello callback. This is the basic building block
+// by which we build a hello-capturing server.
+type capturingConn struct {
+	// Wraps a TCP connection.
+	net.Conn
+
+	helloRead bool
+	helloBuf  *bytes.Buffer
+	helloLock sync.Mutex // protects fields in this block
+
+	onHello onHello
+	onClose func()
+}
+
+func newCapturingConn(wrapped net.Conn, onHello onHello, onClose func()) *capturingConn {
```

Comment:
This method isn't necessary - all the below (other than the hooks) are default values.

Reply:
That's correct - it's a convenience function.  I dunno, I don't feel that strongly about it, but I'm going to leave it as is.